components/style/properties/longhand/box.mako.rs

```diff
@@ -13,229 +13,42 @@
-    #[derive(Clone, Copy, Debug, Eq, Hash, MallocSizeOf, PartialEq, ToComputedValue)]
-    #[cfg_attr(feature = "servo", derive(Deserialize, Serialize))]
-    pub enum SpecifiedValue {
-        % for value in values:
-            ${to_camel_case(value)},
-        % endfor
-    }
-
-    // TODO(emilio): derive.
-    impl ToCss for SpecifiedValue {
-        fn to_css<W>(&self, dest: &mut W) -> ::std::fmt::Result
-            where W: ::std::fmt::Write,
-        {
-            match *self {
-                % for value in values:
-                    SpecifiedValue::${to_camel_case(value)} => dest.write_str("${value}"),
-                % endfor
-            }
-        }
-    }
-
-    /// The initial display value.
-    #[inline]
-    pub fn get_initial_value() -> computed_value::T {
-        computed_value::T::${to_camel_case(values[0])}
-    }
-
-    /// Parse a display value.
-    pub fn parse<'i, 't>(_context: &ParserContext, input: &mut Parser<'i, 't>)
-                         -> Result<SpecifiedValue, ParseError<'i>> {
-        try_match_ident_ignore_ascii_case! { input,
-            % for value in values:
-                "${value}" => {
-                    Ok(computed_value::T::${to_camel_case(value)})
-                },
-            % endfor
-            % for value in webkit_prefixed_values:
-                "-webkit-${value}" => {
-                    Ok(computed_value::T::${to_camel_case(value)})
-                },
-            % endfor
-        }
-    }
-
-    % if product == "servo":
-        fn cascade_property_custom(_declaration: &PropertyDeclaration,
-                                   context: &mut computed::Context) {
-            longhands::_servo_display_for_hypothetical_box::derive_from_display(context);
-            longhands::_servo_text_decorations_in_effect::derive_from_display(context);
-        }
-    % endif
-
-    ${helpers.gecko_keyword_conversion(Keyword('display', ' '.join(values),
-                                               gecko_enum_prefix='StyleDisplay',
-                                               gecko_strip_moz_prefix=False))}
+${helpers.predefined_type(
+    "display",
+    "Display",
+    "computed::Display::inline()",
+    initial_specified_value="specified::Display::inline()",
+    animation_value_type="discrete",
+    needs_context=False,
+    custom_cascade= product == 'servo',
+    custom_cascade_function="specified::Display::cascade_property_custom",
+    flags="APPLIES_TO_PLACEHOLDER",
+    spec="https://drafts.csswg.org/css-display/#propdef-display",
+)}
 
-</%helpers:longhand>
+// FIXME(emilio): Listing all the display values here is very unfortunate, we should teach C++ to use the
+// Rust enum directly, or generate the conversions to `StyleDisplay`.
+<%
+    display_values = """inline block inline-block
+    table inline-table table-row-group table-header-group table-footer-group
+    table-row table-column-group table-column table-cell table-caption
+    list-item none
```

Comment:
This is Gecko-only code now, so no need to ad the `if product == "gecko":`, nor special-casing `webkit_prefixed_display_values`, just put them on the same list.

Reply:
Yeah! You're right!!
I'll put the string of `display_values` as the second argument of `Keyword` and remove this variable.